apps/prairielearn/src/lib/instructorFiles.js

```diff
@@ -7,166 +7,188 @@ import { contains } from '@prairielearn/path-utils';
 
 import { encodePath } from './uri-util.js';
 
+/** @typedef {ReturnType<typeof getPaths>} Paths */
+
 /**
  * For the file path of the current page, this function returns rich
  * information about higher folders up to a certain level determined by
  * the navPage. Created for use in instructor file views.
  *
- * @param {import('express').Request} req
- * @param {import('express').Response} res
+ * @param {string | undefined} requestedPath
+ * @param {Record<string, any>} locals
  */
-export function getPaths(req, res) {
-  let paths = {
-    coursePath: res.locals.course.path,
-    courseId: res.locals.course.id,
-  };
+export function getPaths(requestedPath, locals) {
+  /** @type {string} */
+  const coursePath = locals.course.path;
+  /** @type {string} */
+  const courseId = locals.course.id;
 
-  if (res.locals.navPage === 'course_admin') {
-    paths.rootPath = res.locals.course.path;
-    paths.invalidRootPaths = [
-      path.join(paths.rootPath, 'questions'),
-      path.join(paths.rootPath, 'courseInstances'),
-    ];
-    paths.cannotMove = [path.join(paths.rootPath, 'infoCourse.json')];
-    paths.clientDir = path.join(paths.rootPath, 'clientFilesCourse');
-    paths.serverDir = path.join(paths.rootPath, 'serverFilesCourse');
-    paths.urlPrefix = `${res.locals.urlPrefix}/course_admin`;
-  } else if (res.locals.navPage === 'instance_admin') {
-    paths.rootPath = path.join(
-      res.locals.course.path,
-      'courseInstances',
-      res.locals.course_instance.short_name,
-    );
-    paths.invalidRootPaths = [path.join(paths.rootPath, 'assessments')];
-    paths.cannotMove = [path.join(paths.rootPath, 'infoCourseInstance.json')];
-    paths.clientDir = path.join(paths.rootPath, 'clientFilesCourseInstance');
-    paths.serverDir = path.join(paths.rootPath, 'serverFilesCourseInstance');
-    paths.urlPrefix = `${res.locals.urlPrefix}/instance_admin`;
-  } else if (res.locals.navPage === 'assessment') {
-    paths.rootPath = path.join(
-      res.locals.course.path,
+  /** @type {string} */
+  let rootPath;
+  /** @type {string[]} */
+  let invalidRootPaths;
+  /** @type {string[]} */
+  let cannotMove;
+  /** @type {string} */
+  let clientDir;
+  /** @type {string} */
+  let serverDir;
+  /** @type {string|undefined} */
+  let testsDir;
+  /** @type {string} */
+  let urlPrefix;
+  if (locals.navPage === 'course_admin') {
+    rootPath = coursePath;
+    invalidRootPaths = [path.join(rootPath, 'questions'), path.join(rootPath, 'courseInstances')];
+    cannotMove = [path.join(rootPath, 'infoCourse.json')];
+    clientDir = path.join(rootPath, 'clientFilesCourse');
+    serverDir = path.join(rootPath, 'serverFilesCourse');
+    urlPrefix = `${locals.urlPrefix}/course_admin`;
+  } else if (locals.navPage === 'instance_admin') {
```

Comment:
Consider moving all this to another function like `getPathInfo` so that we can use early returns to (IMO) clean things up a bit:

```ts
function getPathInfo(requestedPath, locals) {
  if (locals.navPage === 'course_admin') {
    return {
      rootPath: coursePath,
      // ...
    }
  }

  if (locals.navPage === 'assessment') {
    // ...
  }

  throw new Error('...');
}
```

Then, in `getPaths`:

```ts
const { rootPath, invalidRootPaths, cannotMove, ... } = getPathInfo(...);
```

Reply:
Done in ed93126c6.